### ISSUE TITLE
add `FlopCounterMode` documentation

### ISSUE THREAD

**Issue by stas00** [pytorch/pytorch#123800]
### 📚 The doc issue

requesting to add `FlopCounterMode` documentation

and to include notes from these discussion:
- https://pytorch.slack.com/archives/C3PDTEV8E/p1712785539175899 (elementwise ops aren't counted - and why)
- https://pytorch.slack.com/archives/C3PDTEV8E/p1712798211869749?thread_ts=1712786199.970369&cid=C3PDTEV8E  torch.compile usage: the intended usage is to use it on an uncompiled version of the model first.

### Suggest a potential alternative/fix

tag: @Chillee

**Comment by alexsamardzic** [comment]
This is going to be hard to get right...  Below is an example where even for matmul the number reported will be wrong: in the case of using 2:4 sparsity, sparse operand will still report the shape of original matrix, while it actually has twice less columns than this matrix, thus the number reported by FlopCounter is twice higher than the actual flop number.  This particular example is indeed rather specific, also it kind of goes under "custom kernel" label, but still - the same problem is for matmul with any kind of sparsity involved.

Furthermore, not all flops are equal, so using the number reported as a performance measure is questionable.  For example, in the `torch.matmul(a, b) * 2` calculation mentioned on Slack, the matmul part will be executed on tensor cores, while element-wise multiplication will be executed on regular CUDA cores, and these flops are not the same w.r.t. throughput (flop/s).

<details>
 <summary>An example with wrong flops number reported for matmul (note: will run on 8.x GPUs only)</summary>

```Python
import torch

from torch.utils.flop_counter import FlopCounterMode

from torch.sparse import to_sparse_semi_structured, SparseSemiStructuredTensor

m, n, p = 64, 64, 64

dtype = torch.float16
device = "cuda"
mask = torch.tensor([0, 0, 1, 1], dtype=torch.bool, device=device).tile(m, n // 4)
a = torch.randn((m, n), dtype=dtype, device=device) * mask.to(dtype)
b = torch.randn((n, p), dtype=dtype, device=device)

asp = to_sparse_semi_structured(a)
flop_counter = FlopCounterMode(display=True)
with flop_counter:
    c = torch.mm(asp, b)
total_flops = flop_counter.get_total_flops()
expected_flops = m * n * p
print("dynamic: ", total_flops)
print("expected: ", expected_flops)
```
</details>

**Comment by carmocca** [comment]
One more ask from my side is to add an example of how to register formulas or specific values for operators that are not supported. For instance, `torch.where(condition)` fails as it is data-dependent, but allowing the user to register a fixed number of flops associated to it (because the user might know the average or worst case shapes) would be good enough to continue with the rest of the total sum for the model.

**Comment by stas00** [comment]
Thank you for the insights, @alexsamardzic!

The currently used manual flos estimator counter functions will not differentiate between tensor cores vs regular cores either. But they could be adapted to take into account sparsity.

So clearly this too might not be suitable for comparing performance of 2 unrelated systems, but it'd work perfectly well for the same system for observing improvements and degradations in performance. In a way it becomes more akin to iter-per-sec counters, which can't be compared between 2 unrelated systems, with the bonus that here you will be given a number that you can still correlate to the theoretical TFLOPS of a dtype, whereas with iter-per-sec counter - the number is completely meaningless when it stands alone.

Bottom line - this tool will be useful for comparing this-apple-to-itself and not apples-to-apples.

**Comment by Chillee** [comment]
> This is going to be hard to get right... Below is an example where even for matmul the number reported will be wrong: in the case of using 2:4 sparsity, sparse operand will still report the shape of original matrix, while it actually has twice less columns than this matrix, thus the number reported by FlopCounter is twice higher than the actual flop number. 

I actually disagree that this means it's "hard to get right". I think it just means that we have specifically defined semantics. For example, Nvidia literally reports their 2:4 sparsity numbers as having 2x the FLOPS :P 

In essence, this FlopCounter is basically counting "theoretical FLOPS of a model using the naive FLOP count for each operator".

Of course, there are many cases where this does not correspond to actual performance. Like you mentioned, non-tensor core FLOPs != tensor-core FLOPs. But it is still useful as a roofline metric to correspond to.

**Comment by alexsamardzic** [comment]
> I actually disagree that this means it's "hard to get right". I think it just means that we have specifically defined semantics. For example, Nvidia literally reports their 2:4 sparsity numbers as having 2x the FLOPS :P

That's true, but then FLOPS numbers reported by NVIDIA depend on data-type too: matmul on F16 tensors has peak FLOPS reported twice bigger than matmul on TF32 tensors, and alike - does it mean FlopCounter should take these into account too?   There are number of other quirks, for example: peak FLOPS on tensor cores is oftentimes hard to achieve without trying number of points in the parameters space (by parameters, I mean on size of tiles handled by blocks/warps of threads, number of "stages" in a software pipeline that is typically used within the matmul kernel, etc.) - this is why we have auto-tuning in Torch compiler, that could help performance a lot.  By "hard to get right", I meant primarily "hard to correlate to actual performance numbers", that some (many?) users of FlopCounter may imply.  Thus, indeed it would be good to stress on the actual semantics of FlopCounter reported numbers in the docs.

**Comment by Chillee** [comment]
I don't see why we would report a different number for F16 tensors vs TF32 tensors. We're reporting FLOPs (i.e. the number of floating point operations performed). This is a well-defined quantity.

**Comment by alexsamardzic** [comment]
I'm not saying we should, was asking following along the argument that for 2:4 sparsity may be OK to report twice more FLOPs than the actual number, partly because of that's how peak FLOPS number for this operation is advertised by NVIDIA...  Anyway, it's really not my intention to nitpick here, just putting my vote for FlopCounter semantics to be stated clearly, as much as possible, in the docs.

**Comment by stas00** [comment]
And the more we discuss this the more I feel that the current name is misleading and the PyTorch team will have many Issues posted in the future asking about wrong outcomes and disappointments. A different name that tells the user what it really can count would produce less friction in the future.

Something like `HighEndPureTorchOPApproximateFlopCounter`... or `SelectiveFlopCounter`?

**Comment by Chillee** [comment]
In this case, I don't think I agree. If users are expecting something different from a flop counter, that's their problem :)

**Comment by stas00** [comment]
All I can share is that I felt a disappointment once I tried this tool, since I hoped it'd give me a magical solution and not require me to painstakingly write out a manual estimator and making sure that they aren't missing some accounting.

I'd still use it as a secondary counter, to be used for relative performance improvements (or detecting regression), but alas, the manual estimator isn't going to go away since we need it to tell how far are we from the optimum and whether there is a point to continue optimizing or not.

So I'm appreciating the addition of this tool (as many other of your awesome creations), but its name lead me to build a false hope that couldn't be fulfilled.

**Comment by Chillee** [comment]
> require me to painstakingly write out manual estimators and making sure that they aren't missing some accounting.

IMO, it should allow you to not need to write out manual estimators. At least, I haven't seen any manual estimators that folks write that are more accurate than this tool. All the complaints about things this tool doesn't account for are also things that aren't accounted for in manual estimators (non-matmul FLOPs, fp16 vs fp32, etc.)

**Comment by stas00** [comment]
That's true with the exception of custom kernels. And when I load a random model from HF, I have no idea if it uses custom kernels or not. So a manual estimator will cover all bases, but this tool won't.

e.g. some models will conditionally load and use `apex` if it's available and fall back to a slower implementation if it's not, which then would result in this tool giving different results for the mathematically the same modelling code.

Megatron-LM is another suspect where it has quite a few custom kernels. I think Deepspeed as well in some situations.

**Comment by stas00** [comment]
Many moons have passed and `FlopCounterMode` is still an undocumented feature.

Thank you.

**Comment by MufanQiu** [comment]
I found some tutorial introduced this tool but there is no documentation about how to further use this tool and what is its limitations. It would be beneficial that we have one, many thanks!

**Comment by albanD** [comment]
I don't know if @Chillee would have some time for this one?

**Comment by malfet** [comment]
Unassigning @Chillee (<jokingly> as he obviously failed or strict SLO on closing hi-pri issues :P )

But keeping triage review for @albanD to explain the hi-pri label

**Comment by albanD** [comment]
High priority because this is a more and more widely used features and it has strong opinions that can be surprising to users. We should make sure it is documented so it doesn't become a common pitfall.

**Comment by ad8e** [comment]
Also note that the flop counter adds significant overhead - even if you run it on one step and leave it off afterward. It increases my step time from 0.34 sec to 0.40 sec, possibly from extra CPU overhead.

**Comment by zou3519** [comment]
> Also note that the flop counter adds significant overhead even if you run it on one step and leave it off afterward. It increases my step time from 0.34 sec to 0.40 sec, possibly from extra CPU overhead.

@ad8e do you have a repro for this? I will try to come up with one myself, but this is unexpected to me. FlopCounterMode shouldn't be making the rest of training slower if you leave it off after

**Comment by ad8e** [comment]
It should repro in my codebase, so I'll see if my AI can produce a minimal repro.

**Comment by ad8e** [comment]
mahouko's explanation was:
> flop counter is a context manager, it doesn't modify or reassign functions. it adds an observer to operation dispatches
> in that situation, you are entering DispatchMode at the same time as you are telling torch to trace operations for the first time in that model instance’s compiled forward
```
"""Show that torch.compile stays slow after its first call under FlopCounterMode.

Run: python repro_flop_counter_compile_slowdown.py
The ``found no compiled frames`` warning is expected.
"""

import torch
from torch.utils.flop_counter import FlopCounterMode

STEPS = 100


def outside_workload(x, weight):
    x = x @ weight
    for _ in range(25):
        x = torch.sin(x) + 0.01 * x
    return x


def inside_workload(x, weight):
    x = x @ weight
    for _ in range(25):
        x = torch.sin(x) + 0.01 * x
    return x


def time_ms(fn, x, weight):
    start = torch.cuda.Event(enable_timing=True)
    end = torch.cuda.Event(enable_timing=True)
    start.record()
    for _ in range(STEPS):
        fn(x, weight)
    end.record()
    end.synchronize()
    return start.elapsed_time(end) / STEPS


def main():
    if not torch.cuda.is_available():
        raise RuntimeError("This reproduction requires CUDA")

    x = torch.randn(8192, 256, device="cuda", dtype=torch.bfloat16)
    weight = torch.randn(256, 256, device="cuda", dtype=torch.bfloat16)

    compiled_outside = torch.compile(outside_workload, fullgraph=True)
    compiled_outside(x, weight)

    with FlopCounterMode(display=False):
        compiled_inside = torch.compile(inside_workload, fullgraph=True)
        compiled_inside(x, weight)

    # warmup. Every call below happens after FlopCounterMode has exited.
    for _ in range(5):
        compiled_outside(x, weight)
        compiled_inside(x, weight)
    torch.cuda.synchronize()

    outside_ms = time_ms(compiled_outside, x, weight)
    inside_ms = time_ms(compiled_inside, x, weight)
    print(f"compiled outside counter: {outside_ms:.3f} ms/step")
    print(f"compiled inside counter:  {inside_ms:.3f} ms/step")
    print(f"slowdown after exit:      {inside_ms / outside_ms:.2f}x")


if __name__ == "__main__":
    main()
```

Torch 2.12, B200, cuda 13.2. Output:

```
kevin@kevin-b200-0:/mnt/clusterstorage/workspace/kevin$ python repro_flop_counter_compile_slowdown.py
/usr/local/lib/python3.12/dist-packages/torch/_dynamo/eval_frame.py:1069: UserWarning: torch.compile with fullgraph=True found no compiled frames. The frame was likely skipped (e.g., a non-infra torch dispatch mode was active, dynamo was disabled, or the frame was skipped. This may become an error in the future.
  torch._dynamo.utils.warn_once(
compiled outside counter: 0.057 ms/step
compiled inside counter:  0.404 ms/step
slowdown after exit:      7.15x
```

**Comment by zou3519** [comment]
@ad8e thank you for the detailed repro, I opened a dedicated issue for this one: https://github.com/pytorch/pytorch/issues/190286

**Comment by zou3519** [comment]
@ad8e to check: do you have any repros of this (FlopCounterMode slowing down future iterations) without torch.compile? I think you should only see a problem with torch.compile (because that special-cased this situation in an unfortunate way)

**Comment by ad8e** [comment]
Nope, I think torch.compile is the only issue. I bisected our separate slowdown with flop-counter-eager-on-one-step, and it was caused by an internal bug where we cached flop counts for padded shapes but retrieved flop counts for unpadded shapes. So you already solved the main problem with FlopCounter.
